src/odemis/driver/keysight.py

```diff
@@ -0,0 +1,333 @@
+        for m in msgs[:-1]:
+            self._parseMessage(m)  # will update _output_buf
+
+        self._input_buf = msgs[-1]
+
+    def read(self, size=1):
+        ret = self._output_buf[:size]
+        self._output_buf = self._output_buf[len(ret):]
+
+        if len(ret) < size:
+            # simulate timeout
+            time.sleep(self.timeout)
+        return ret
+
+    def close(self):
+        # using read or write will fail after that
+        del self._output_buf
+        del self._input_buf
+
+    def _sendAnswer(self, ans):
+        self._output_buf += "%s\n" % (ans,)
+
+    def _parseMessage(self, msg):
+        """
+        msg (str): the message to parse
+        return None: self._output_buf is updated if necessary
+        """
+        logging.debug("SIM: parsing %s", to_str_escape(msg))
+        msg = msg.decode("latin1").strip()  # remove leading and trailing whitespace
+        msg = "".join(msg.split())  # remove all space characters
+
+        if msg == "*IDN?":
+            self._sendAnswer("Agilent Technologies,33622A,MY59003187,A.02.03-3.15-03-64-0")
+        elif re.match(r"sour[1-2]:freq ", msg):
+            channel = msg[4]
+            period = 1 / float(msg.split()[1])
+            self._output_states['channel' + channel]['period'] = period
+        elif re.match(r"sour[1-2]:trac ", msg):
+            channel = msg[4]
+            tracking_mode = msg.split()[1]
+            self._output_states['channel' + channel]['tracking'] = tracking_mode
+        elif re.match(r"sour[1-2]:volt:low ", msg):
+            channel = msg[4]
+            voltage_low = float(msg.split()[1])
+            self._output_states['channel' + channel]['voltage_low'] = voltage_low
+        elif re.match(r"sour[1-2]:volt:high ", msg):
+            channel = msg[4]
+            voltage_high = float(msg.split()[1])
+            self._output_states['channel' + channel]['voltage_high'] = voltage_high
+        elif re.match(r"sour[1-2]:volt:offs ", msg):
+            channel = msg[4]
+            voltage_offset = float(msg.split()[1])
+            self._output_states['channel' + channel]['voltage_offset'] = voltage_offset
+        elif re.match(r"sour[1-2]:volt:lim:high ", msg):
+            channel = msg[4]
+            voltage_lim_min = float(msg.split()[1])
+            self._output_states['channel' + channel]['voltage_limit_low'] = voltage_lim_min
+        elif re.match(r"sour[1-2]:volt:lim:low ", msg):
+            channel = msg[4]
+            voltage_lim_max = float(msg.split()[1])
+            self._output_states['channel' + channel]['voltage_limit_high'] = voltage_lim_max
+        elif re.match(r"sour[1-2]:func:squ:dcyc ", msg):
+            channel = msg[4]
+            duty_cycle = float(msg.split()[1])
+            # duty cycle value is between 0 and 1
+            self._output_states['channel' + channel]['duty_cycle'] = float(duty_cycle / 100)
+
+        else:
+            logging.info("Invalid command: %s", msg)
```

Comment:
That should be higher than info, as it's a sign something is gone really unexpected. I'd say "error" the the proper level.
Do you know how the device react in such case? Does it return a specific value? Or just ignore the command? Do you have this in your logs?

Reply:
It does not return a specific value and times out eventually(so returns timeout error).

Additionally on the hardware screen it shows: `"Error generated by remote interface"` error message.